server/utils/recipeUtils.js

```diff
@@ -46,8 +47,71 @@ const getByID = asyncHandler(async (recipeID, Recipe) => {
     return recipe
 })
 
+const getShoppingList = asyncHandler(async (userId, Recipe, Ingredient) => {
+	await Ingredient.sync()
+	Recipe.hasMany(Ingredient, {foreignKey: 'recipe_id'})
+	Ingredient.belongsTo(Recipe, {foreignKey: 'recipe_id'})
+	console.log(Ingredient);
+	const recipes = await Ingredient.findAll({
+		attributes: [
+			'ingredient_name', 
+			'ingredient_unit', 
+			[
+				Sequelize.literal('SUM(Recipe.portion*ingredient_amount)'), 
+				'total_amount'
+			]
+		], 
+		include: [
+			{
+				model: Recipe, 
+				required: true, 
+				attributes: [], 
+				where: {
+					user_id: userId, 
+					portion: {[Op.gt]: "0"}
+				}
+			}
+		], 
+		group: [
+			'ingredient_name', 
+			'ingredient_unit'
+		], 
+		raw: true
+	});
+	return recipes
+})
+
+const setPortion = asyncHandler(async (recipeID, portions, userId, Recipe) => {
+	await Recipe.sync()
+	const recipe = await Recipe.findOne({ where: {id: recipeID, user_id: userId}});
+	
+	if(recipe){
+
+		recipe.portion = portions
+
+		recipe.save()
+		return true;
+	}
+	else{
+		return false;
+	}
+})
+
+const resetPortions = asyncHandler(async (userId, Recipe) => {
+	await Recipe.sync()
+	const recipes = await Recipe.findAll({ where: {user_id: userId}})
+
+	for(i=0; i<recipes.length; i++){
+		recipes[i].portion = 0;
+		recipes[i].save();
```

Comment:
In addition to resetting the portion to 0, we also want to check if the recipe is in the shopping list (param 'selected' in recipeModel is correctly set to True) before remove it from the shopping list (change the param to False). Similarly, when setting a portion size to 0, this param also needs to be adjusted.

Reply:
Do we need to use the `selected` param? Since if `portion > 0` that means that it should be included in the shopping list and if `portion = 0`, the recipe shouldn't be included in the shopping list. So we can derive whether the recipe is included in the shopping list just by the `portion` value alone

Reply:
Agreed, will remove this param from recipeModel on my next PR.